File: gles-library/src/main/cpp/render/GLRenderer.cpp
//
// Created by gtbluesky on 2019-07-18.
//

#include "GLRenderer.h"
#include "../shape/Triangle.h"

GLRenderer::~GLRenderer() {
    if (mEglCore) {
        mEglCore->release();
        delete mEglCore;
        mEglCore = nullptr;
    }
}

void GLRenderer::surfaceCreated(ANativeWindow *window) {
    if (!mEglCore) {
        mEglCore = new EglCore(nullptr, FLAG_RECORDABLE);
    }
    mWindowSurface = new WindowSurface(mEglCore, window, false);
    mWindowSurface->makeCurrent();
    shape = new Triangle();
    shape->init();
}

void GLRenderer::surfaceChanged(int width, int height) {
    mWindowSurface->makeCurrent();
    shape->onDraw(width, height);
    mWindowSurface->swapBuffers();
}

void GLRenderer::surfaceDestroyed() {
    if (shape) {
        shape->destroy();
        delete shape;
        shape = nullptr;
    }
    if (mWindowSurface) {
        mWindowSurface->release();
        delete mWindowSurface;
        mWindowSurface = nullptr;
    }
    if (mEglCore) {
        mEglCore->release();
        delete mEglCore;
        mEglCore = nullptr;
    }
}

File: gles-lib/src/main/java/com/gtbluesky/gles/render/GLRenderer.kt
package com.gtbluesky.gles.render

import android.view.Surface
import com.gtbluesky.gles.egl.EglCore
import com.gtbluesky.gles.egl.WindowSurface
import com.gtbluesky.gles.shape.BaseShape
import com.gtbluesky.gles.shape.Cube
import com.gtbluesky.gles.shape.Sphere

class GLRenderer {

    private var eglCore: EglCore? = null
    private var windowSurface: WindowSurface? = null
    private var shape: BaseShape? = null

    fun init(surface: Surface, shape: BaseShape = Cube()) {
        eglCore = eglCore ?: EglCore(
            null,
            EglCore.FLAG_RECORDABLE
        )
        eglCore?.let {
            windowSurface = WindowSurface(it, surface, false)
            windowSurface?.makeCurrent()
        }
        this.shape = shape
        this.shape?.init()
    }

    fun change(width: Int, height: Int) {
        windowSurface?.makeCurrent()
        shape?.change(width, height)
    }

    fun drawFrame() {
        shape?.drawFrame()
        windowSurface?.swapBuffers()
    }

    fun destroy() {
        shape?.destroy()
        windowSurface?.release()
        eglCore?.release()
        shape = null
        windowSurface = null
        eglCore = null
    }
}